From: Michael Olbrich <m.olbrich@pengutronix.de>
Date: Mon, 18 Mar 2019 10:42:15 +0100
Subject: [PATCH] don't add rpaths for build directories when cross-compiling

The binaries cannot be executed anyways, so this just increases the binary
size unnecessarily.

Instead add -rpath-link so ensure the libraries are found at build-time.

Signed-off-by: Michael Olbrich <m.olbrich@pengutronix.de>
---
 mesonbuild/linkers/linkers.py | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/mesonbuild/linkers/linkers.py b/mesonbuild/linkers/linkers.py
index d40d5e20effe..35b77143c1cb 100644
--- a/mesonbuild/linkers/linkers.py
+++ b/mesonbuild/linkers/linkers.py
@@ -756,7 +756,12 @@ class GnuLikeDynamicLinkerMixin(DynamicLinkerBase):
         # Need to deduplicate rpaths, as macOS's install_name_tool
         # is *very* allergic to duplicate -delete_rpath arguments
         # when calling depfixer on installation.
-        all_paths = mesonlib.OrderedSet([os.path.join(origin_placeholder, p) for p in processed_rpaths])
+        if not self.environment.is_cross_build():
+            all_paths = mesonlib.OrderedSet([os.path.join(origin_placeholder, p) for p in processed_rpaths])
+        else:
+            all_paths = mesonlib.OrderedSet()
+            for p in rpath_paths:
+                args.extend(self._apply_prefix('-rpath-link,' + os.path.join(build_dir, p)))
         rpath_dirs_to_remove: T.Set[bytes] = set()
         for p in all_paths:
             rpath_dirs_to_remove.add(p.encode('utf8'))
@@ -788,7 +793,8 @@ class GnuLikeDynamicLinkerMixin(DynamicLinkerBase):
                 paths = padding
             else:
                 paths = paths + ':' + padding
-        args.extend(self._apply_prefix('-rpath,' + paths))
+        if paths:
+            args.extend(self._apply_prefix('-rpath,' + paths))
 
         # TODO: should this actually be "for solaris/sunos"?
         # NOTE: Remove the zigcc check once zig support "-rpath-link"
